wen-restart/src/wen_restart.rs

```diff
@@ -0,0 +1,150 @@
+//! The `wen-restart` module handles automatically repair in cluster restart
+
+use {
+    crate::solana::wen_restart_proto,
+    log::*,
+    prost::Message,
+    solana_gossip::cluster_info::ClusterInfo,
+    solana_ledger::blockstore::Blockstore,
+    solana_vote_program::vote_state::VoteTransaction,
+    std::{
+        fs::File,
+        io::{Error, Write},
+        path::PathBuf,
+        sync::Arc,
+    },
+};
+
+// The number of ancestor slots sent is hard coded at 81000, because that's
+// 400ms * 81000 = 9 hours, we assume most restart decisions to be made in 9
+// hours.
+const MAX_SLOTS_ON_VOTED_FORKS: u32 = 81000;
+
+pub fn wait_for_wen_restart(
+    wen_restart_path: &PathBuf,
+    last_vote: VoteTransaction,
+    blockstore: Arc<Blockstore>,
+    cluster_info: Arc<ClusterInfo>,
+) -> Result<(), Box<dyn std::error::Error>> {
+    // repair and restart option does not work without last voted slot.
+    let last_vote_slot = last_vote.last_voted_slot().unwrap();
+    let mut last_vote_fork = vec![last_vote_slot];
+    let mut slot = last_vote_slot;
+    for _ in 0..MAX_SLOTS_ON_VOTED_FORKS {
```

Comment:
do we need to go this deep, can we stop at `block_commitment_cache.highest_supermajority_root()` ? 

if i'm behind the smr my last voted fork info doesn't contribute anything to wen restart anyway right?

Reply:
That is also an option, I think the current logic is trying to be more straightforward so more fault tolerant. What if there is a bug in code so highest_supermajority_root is wrong?

So when we started the initial design we tried to see what the minimum required info is, we had to send last_vote, and need to send the slots on last voted fork so validators with different progress can sync up again. Other than that, we know an outage is probably less than 9 hours before we intervene, and we avoid relying on all other information. Because software can have bugs, you can't predict what is going wrong in an outage.

Does that make sense?

Reply:
do you know what the latest oom time is, can we actually survive 9 hours without making a root?

either way it seems alright to play it safe and send all these slots. 
was selfishly hoping we could reduce the size so that we could include hashes for the inevitable duplicate block restart 😛